configs/configupgrade/upgrade_expr.go

```diff
@@ -557,10 +557,32 @@ Value:
 		diags = diags.Append(exprDiags)
 		keySrc, exprDiags := upgradeExpr(tv.Key, filename, true, an)
 		diags = diags.Append(exprDiags)
-		buf.Write(targetSrc)
-		buf.WriteString("[")
-		buf.Write(keySrc)
-		buf.WriteString("]")
+
+		parts := strings.Split(string(targetSrc), ".")
+		parts = upgradeTraversalParts(parts, an) // might add/remove/change parts
```

Comment:
`upgradeTraversalParts` is expecting to receive a HIL-style traversal string, but `targetSrc` here is an arbitrary HCL 2.0 expression buffer instead, so I worry this will behave strangely if `upgradeExpr(tv.Target, ...)` has applied some other rewriting already that makes this not be a valid HIL-style traversal string anymore.

In practice I think limitations of the HIL parser guarantee that `tv.Target` is always a `*hilast.VariableAccess` but the upgrade steps for that already include replacement of certain traversal steps with constructs that are not valid HIL traversal strings, such as introducing indexing brackets. For example, if the variable access were `foo.1bar.*.baz` then `upgradeExpr` would rewrite it to `foo["1bar"].*.baz` (because `1bar` is not a valid HCL 2 identifier), which would then split up in a way that `upgradeTraversalParts` is not prepared to deal with.

I think to do this correctly we'd need to check up front whether `tv.Target` is a `*hilast.VariableAccess` and, if so, use its `Name` field value to produce the input to `upgradeTraversalParts`, and then do some behavior similar to what the `*hilast.VariableAccess` rule above does to turn the resulting parts into HCL 2 syntax. I expect that's technically possible with some refactoring to extract those rules into a separate function, but I wonder if the complexity of all that outweighs the value of what is, as far as I can tell, a totally optional rewrite.

Reply:
I did a bit of refactoring and also raised https://github.com/hashicorp/terraform/pull/21125 some parts of which I'm leveraging here.